File: haskell-server/src/BoardGame/Common/Domain/Player.hs
--
-- Copyright 2017 Azad Bolour
-- Licensed under GNU Affero General Public License v3.0 -
--   https://github.com/azadbolour/boardgame/blob/master/LICENSE.md
--

{-# LANGUAGE DeriveGeneric #-}

module BoardGame.Common.Domain.Player (
    PlayerName
  , PlayerType (..)
  , userIndex, machineIndex
  , playerTypeIndex
  , Player(..)
) where

import GHC.Generics
import Data.Aeson

-- | The name of a player.
type PlayerName = String

-- | Types of players.
data PlayerType = UserPlayer | MachinePlayer
  deriving (Show, Read, Eq, Generic)

instance FromJSON PlayerType
instance ToJSON PlayerType

-- | Index of user player - for indexing into an array of user/machine.
userIndex = 0 :: Int
-- | Index of machine player - for indexing into an array of user/machine.
machineIndex = 1 :: Int

-- | Get the index of a player for indexing into an array of user/machine.
playerTypeIndex :: PlayerType -> Int
playerTypeIndex UserPlayer = userIndex
playerTypeIndex MachinePlayer = machineIndex

-- | A user player.
data Player = Player {
    name :: PlayerName    -- ^ Unique name of the player.
}
  deriving (Eq, Show, Generic)

instance FromJSON Player
instance ToJSON Player

File: scala-server/app/com/bolour/boardgame/scala/server/domain/Player.scala
/*
 * Copyright 2017-2018 Azad Bolour
 * Licensed under GNU Affero General Public License v3.0 -
 *    https://github.com/azadbolour/boardgame/blob/master/LICENSE.md
 */
package com.bolour.boardgame.scala.server.domain

import com.bolour.util.scala.common.CommonUtil.{ID, Email}
import com.bolour.util.scala.server.BasicServerUtil.{stringId}

/**
  * Player.
  *
  * @param id The unique identifier of the player in the board game database.
  * @param userId The unique identifier of the player as a user
  *                       in the authentication provider's system.
  * @param name The name of the user. For maximum security we just use the first name.
  * @param email The player's email.
  *              The email is the "natural key" of the user in the boardgame system.
  */
case class Player(id: ID, userId: String, name: String, email: Email)

